packages/grid/_modules_/grid/hooks/features/selection/useGridSelection.ts

```diff
@@ -67,30 +71,29 @@ export const useGridSelection = (apiRef: GridApiRef): void => {
 
       logger.debug(`Selecting row ${id}`);
 
-      const allowMultiSelect =
-        allowMultipleOverride || allowMultipleSelectionKeyPressed.current || checkboxSelection;
+      setGridState((state) => {
+        let selectionState: GridSelectionState = { ...state.selection };
+        const allowMultiSelect =
+          allowMultipleOverride ||
+          (!disableMultipleSelection && isMultipleKey) ||
+          checkboxSelection;
 
-      if (allowMultiSelect) {
-        setGridState((state) => {
-          // eslint-disable-next-line prefer-object-spread
-          const selectionState: GridSelectionState = Object.assign({}, state.selection);
+        if (allowMultiSelect) {
           const isRowSelected =
             !allowMultiSelect || isSelected == null ? selectionState[id] === undefined : isSelected;
```

Comment:
`allowMultiSelect` is always true in this branch. 😛

```suggestion
            isSelected == null ? selectionState[id] === undefined : isSelected;
```